### PR TITLE
Dtlz problems

### PR DESCRIPTION
**Pull request by darioizzo**
This PR ports the dtlz problem suite which is needed to write a good tutorial for the hypervolumes

### CODE REVIEW

File: include/pagmo/problems/dtlz.hpp [hunk 1]
```diff
@@ -0,0 +1,511 @@
...
+You should have received copies of the GNU General Public License and the
+GNU Lesser General Public License along with the PaGMO library.  If not,
+see https://www.gnu.org/licenses/. */
+
+#ifndef PAGMO_PROBLEM_DTLZ_HPP
+#define PAGMO_PROBLEM_DTLZ_HPP
+
+#include <algorithm>
+#include <cmath>
+#include <exception>
```
Comment (bluescarni) [inline include/pagmo/problems/dtlz.hpp]: You probably mean ``#include <stdexcept>``.

File: include/pagmo/problems/dtlz.hpp [hunk 2]
```diff
@@ -0,0 +1,511 @@
...
+#include "../population.hpp"
+#include "../problem.hpp"
+#include "../types.hpp"
+
+namespace pagmo
+{
+/// DTLZ problem test suite
+/**
+ * This widespread test suite was conceived for multiobjective problems with scalable fitness dimensions
+ * and takes its name from its authors Deb, Thiele, Laumanns and Zitzler
```
Comment (bluescarni) [inline include/pagmo/problems/dtlz.hpp]: Missing ``.``

File: include/pagmo/problems/dtlz.hpp [hunk 3]
```diff
@@ -0,0 +1,511 @@
...
+ *
+ * All problems in this test suite are box-constrained continuous n-dimensional multi-objective problems, scalable in
+ * fitness dimension. The dimension of the decision space is \f$ k + fdim - 1 \f$, whereas fdim is the number of
+ * objectives and k a paramter. Properties of the decision space and the Pareto-front of each problems are as follow:
```
Comment (bluescarni) [inline include/pagmo/problems/dtlz.hpp]: "as follows"

File: include/pagmo/problems/dtlz.hpp [hunk 4]
```diff
@@ -0,0 +1,511 @@
...
+ * DTLZ6:
+ *
+ * A more difficult version of the DTLZ5 problem: the non-linear distance function g makes it harder to convergence
+ * against the pareto optimal curve.
+ *
+ * DTLZ7:
+ *
+ * This problem has disconnected Pareto-optimal regions in the search space.
+ *
+ * @see K. Deb, L. Thiele, M. Laumanns, E. Zitzler, Scalable test problems for evolutionary multiobjective optimization
```
Comment (bluescarni) [inline include/pagmo/problems/dtlz.hpp]: "See:"

File: include/pagmo/problems/dtlz.hpp [hunk 5]
```diff
@@ -0,0 +1,511 @@
...
+                break;
+            case 7:
+                retval = f7_objfun_impl(x);
+                break;
+        }
+        return retval;
+    }
+    /// Number of objectives
+    /**
+     * One of the optional methods of any user-defined problem (UDP).
```
Comment (bluescarni) [inline include/pagmo/problems/dtlz.hpp]: I'd drop this description, just documenting with ``@return`` is enough.

File: include/pagmo/problems/dtlz.hpp [hunk 6]
```diff
@@ -0,0 +1,511 @@
...
+     *
+     * @return the number of objectives
+     */
+    vector_double::size_type get_nobj() const
+    {
+        return m_fdim;
+    }
+    /// Box-bounds
+    /**
+     * One of the optional methods of any user-defined problem (UDP).
```
Comment (bluescarni) [inline include/pagmo/problems/dtlz.hpp]: Here instead there's useful information missing from the docstring: I can't see anywhere documented that the bounds are [0,1] in all dimensions.

In general, we don't need to repeat on each method implemented in the UDP that "this is a method which is part of the UDP interface" etc. etc., we just need to document its behaviour in that specific case. If it's already documented elsewhere (e.g., in the class' description), then ``@param`` and ``@return`` are enough.

File: include/pagmo/problems/dtlz.hpp [hunk 7]
```diff
@@ -0,0 +1,511 @@
...
+     * It returns the box-bounds for this UDP.
+     *
+     * @return the lower and upper bounds for each of the decision vector components
+     */
+    std::pair<vector_double, vector_double> get_bounds() const
+    {
+        std::pair<vector_double, vector_double> retval{vector_double(m_dim, 0.), vector_double(m_dim, 1.)};
```
Comment (bluescarni) [inline include/pagmo/problems/dtlz.hpp]: Maybe compress on a single line with the init list syntax? ``return {vector_double(...), vector_double(...)};``.

File: include/pagmo/problems/dtlz.hpp [hunk 8]
```diff
@@ -0,0 +1,511 @@
...
+    {
+        if (x.size() != m_dim) {
+            pagmo_throw(std::invalid_argument, "The size of the decision vector should be " + std::to_string(m_dim)
+                                                   + " while " + std::to_string(x.size()) + " was detected");
+        }
+        return convergence_metric(x);
+    }
+    /// Problem name
+    /**
+     * One of the optional methods of any user-defined problem (UDP).
```
Comment (bluescarni) [inline include/pagmo/problems/dtlz.hpp]: Ditto.

File: include/pagmo/problems/dtlz.hpp [hunk 9]
```diff
@@ -0,0 +1,511 @@
...
+        // to have the minimum at 0.0 so we can use the p_distance implementation in base_dtlz
+        // to have the p_distance converging towards 0.0 rather then towards 1.0
+        double y = 0.;
+        for (decltype(x.size()) i = 0u; i < x.size(); ++i) {
+            y += x[i];
+        }
+        return (9. / static_cast<double>(x.size())) * y;
+    }
+    /// Implementation of the distribution function h
+    double h7_func(const vector_double &f, const double g) const
```
Comment (bluescarni) [inline include/pagmo/problems/dtlz.hpp]: Probably an ``&`` is missing here.

File: pygmo/test.py [hunk 10]
```diff
@@ -1804,7 +1804,6 @@ class population_test_case(_ut.TestCase):
     """Test case for the :class:`~pygmo.core.population` class.
 
     """
-
```
Comment (bluescarni) [inline pygmo/test.py]: Are these removed spaces an artefact of the editor or is it autopep8 stuff?

Comment (darioizzo) [inline pygmo/test.py]: maybe my new editor settings? mah

**Comment by bluescarni** [comment]
👍

**Pull request merged**